revng-distributable revamp

mrjackv opened the pull request.

mrjackv committed:
Refactor revng-distributable

mrjackv committed:
Remove unused test.sh

aleclearmind commented on .orchestra/ci/install-old-system-dependencies.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Move this as a comment in the `qt` component.

Did you really think a file with `old` in the name was going to pass code review? `:D`

mrjackv replied:
Fixed

aleclearmind commented on .orchestra/support/revng-distributable/test.sh:
```diff
@@ -1,53 +0,0 @@
-#!/bin/bash
```
Can we drop the UI tests but not the others?

mrjackv replied:
The script itself is not used anywhere in orchestra

aleclearmind commented on .orchestra/config/components/revng_distributable.yml:
```diff
@@ -9,113 +9,9 @@ configure: |
-  cd "${TMP_ROOT}${ORCHESTRA_ROOT}"
-  find . -not -type d -not -path './revng/*' -delete
-  find . -type d -empty -delete
+  "$ORCHESTRA_DOTDIR/support/revng-distributable/install.sh"
```
Hash check?

aleclearmind replied:
Updates?

mrjackv replied:
Fixed

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,102 @@
+# Manually add PipelineC headers
+mkdir -p "${TMP_ROOT}${ORCHESTRA_ROOT}/revng/root/include/revng/PipelineC"
+cp -a "$ORCHESTRA_ROOT"/include/revng/PipelineC/*.h \
+      "${TMP_ROOT}${ORCHESTRA_ROOT}/revng/root/include/revng/PipelineC"
```
Please do this using `rsync`

mrjackv replied:
Fixed

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,102 @@
+  "root/bin/"
+
+echo "Copying README.md"
+cp -a "$ORCHESTRA_DOTDIR/support/revng-distributable/README.md" .
```
Review the README, does it still make sense?

mrjackv replied:
Fixed it up, removing mentions of the UI.
However in general it depends on our intentions w.r.t. reusing `revng-distributable` also for local install tarballs or not (I highly doubt anyone will `cat` a file inside a docker container)

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,102 @@
+cp -a "$ORCHESTRA_DOTDIR/support/revng-distributable/README.md" .
+
+echo "Copying install-revng-dependencies"
+cp -a "$ORCHESTRA_DOTDIR/support/revng-distributable/dockers/install-revng-dependencies" root/bin/
```
Are you using this in your `Dockerfile`?

mrjackv replied:
No, I left that section of the installer as-is

aleclearmind commented on .orchestra/config/components/ui/qt.yml:
```diff
@@ -4,6 +4,13 @@
+#! gperf libcap-dev libfontconfig1-dev libfreetype6-dev libgl-dev libgl1-mesa-dev libgles2-mesa-dev
+#! libinput-dev libmount-dev libssl-dev libx11-dev libx11-xcb-dev libxcb-glx0-dev libxcb1-dev
+#! libxext-dev libxfixes-dev libxi-dev libxkbcommon-dev libxkbcommon-x11-dev libxrender-dev
+
```
We really need to get rid of all the components we don't care about.
At least all the Qt ones plus caliban and cold-revng.

I'm wary of dropping the X stuff in case we decide to build VSCode in orchestra, that's still a decision we have to think about.

mrjackv replied:
Orthogonally, I'd say to drop everything and if we ever need to build X-related stuff we can `git checkout` the component yamls. Right now they're there making clutter and doing nothing.

aleclearmind replied:
I'm cleaning up things, will ping you once you can rebase.

aleclearmind commented on .orchestra/support/revng-container/apt.no-extra:
```diff
@@ -0,0 +1,4 @@
+APT::Get::Install-Recommends "false";
+APT::Get::Install-Suggests "false";
+APT::Install-Suggests "0";
+APT::Install-Recommends "0";
```
Missing blank line, in other places too.

mrjackv replied:
Should I `revng-check-convention` the rest of orchesta? there's a lot of stuff that's not passing
(Also there's the catch22 problem of checking orchestra with rcc when orchestra has to install rcc)

aleclearmind replied:
Nah.

aleclearmind commented on .orchestra/config/components/llvm_documentation.yml:
```diff
@@ -56,9 +56,9 @@ builds:
+      cp -far "$BUILD_DIR/build/docs/doxygen/html/llvm.docset" "${DESTDIR}${ORCHESTRA_ROOT}/share/doc/llvm"
+      cp -far "$BUILD_DIR/build/tools/clang/docs/doxygen/html/clang.docset" "${DESTDIR}${ORCHESTRA_ROOT}/share/doc/clang"
+      cp -far "$BUILD_DIR/build/tools/mlir/docs/doxygen/html/mlir.docset" "${DESTDIR}${ORCHESTRA_ROOT}/share/doc/mlir"
 
```
This commit is in another PR.

mrjackv replied:
Dropped

aleclearmind commented on .gitignore:
```diff
@@ -2,7 +2,7 @@ root
 sources
 .orchestra/tmproot
-.orchestra/binary-archives/*
+.orchestra/binary-archives
```
This commit is already in another PR.

mrjackv replied:
Dropped

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,125 @@
+    echo strip "$FILE"
+    strip "$FILE"
+  fi
+done < "$STRIP_FILE_LIST"
```
Do we still get nice stack traces if we do this? If not, we can't do this.

mrjackv replied:
We get line numbers + filename for our libraries
LLVM libraries just get the demangled function name

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,125 @@
+        -e "^share/orchestra/" \
+        -e "^share/revng/pipelines/" \
+        -e "^share/revng/[^/]*\.ll$" \
+        -e "^x86_64-pc-linux-gnu/" | \
```
What criterion did you use to put together this list?
There are no comments and it's very fragile and easy to break.

mrjackv replied:
> What criterion did you use to put together this list?

Gut feelings and checking if the `test.py` passed
I can add comments, but there's no formal way for us to say "this file is needed for runtime"

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,125 @@
+echo "Generating checksums"
+cd "$TMP_ROOT/$ORCHESTRA_ROOT/revng"
+find root -type f -exec sha256sum {} \; > checksums.sha256
+sha256sum README.md environment revng >> checksums.sha256
```
What are these for?

mrjackv replied:
Just a nicety.
So that if you ever receive the raw tarball you can do `sha256sum -c checksums.sha256` and be sure it's all ok

aleclearmind replied:
I was looking for a canonical name for such files, but I can't find anything, so this works.

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,125 @@
+  else
+    rm -f "$IDX" "${IDX//.idx/.license}" "${IDX//.idx/.json}"
+  fi
+done
```
What's going on here?
We can't drop the license files from the release.

mrjackv replied:
Basically it's rewriting the `.idx` files, trimming away any file that has been removed
The `rm` is in the case _all_ the files of the component have been deleted, to keep things tidy

aleclearmind commented on .orchestra/support/distributable-test.py:
```diff
@@ -0,0 +1,188 @@
+            if not target["serialized"].endswith(f":{artifacts_kind}"):
+                continue
+
+            print(f"Producing {step_name}/{artifacts_container}/{target['serialized']}")
```
If it's debug information, make `log` function and write to `stderr`.

mrjackv replied:
Done

aleclearmind commented on .orchestra/support/distributable-test.py:
```diff
@@ -0,0 +1,188 @@
+        for connection in connections:
+            if connection.raddr == () and connection.status == "LISTEN":
+                return connection.laddr.port
+        sleep(0.5)
```
We need an alternative solution to this.

mrjackv replied:
Fixed via unix sockets

aleclearmind commented on .orchestra/support/distributable-test.py:
```diff
@@ -0,0 +1,188 @@
+        sleep(0.5)
+
+
+def main():
```
This belongs to the revng project, or it will be easier to make it go out of sync with revng (e.g., changes int the GraphQL API). Maybe we can have a `revng self-test` command?

mrjackv replied:
I've added in the sister PR for revng the command `revng daemon-self-test`

aleclearmind commented on .orchestra/support/distributable-test.py:
```diff
@@ -0,0 +1,188 @@
+        traceback.print_exception(e.__class__, e, e.__traceback__)
+        process.communicate()
+        print_fd(out_fd)
+        sys.exit(1)
```
Document why this is better than just the exception reach the user.

mrjackv replied:
Dropped

aleclearmind commented on .orchestra/config/components/revng_container.yml:
```diff
@@ -0,0 +1,26 @@
+install: |
+  INSTALLER="$ORCHESTRA_DOTDIR/support/revng-container/install.sh"
+  #echo "bacb4190d2457e2cc2303aed9928f2a3cf33107fe20c0613b68fedf56683f4f9 $INSTALLER" | \
+  #  sha256sum --quiet -c -
```
?

mrjackv replied:
Must have forgot to uncomment and update the hash, fixed

aleclearmind commented on .orchestra/config/components/revng_container.yml:
```diff
@@ -0,0 +1,26 @@
+binary_archives: private
+configure: |
+  mkdir -p "$BUILD_DIR" && cd "$BUILD_DIR"
+  touch empty_directory_ignore_it
```
Is this needed?

mrjackv replied:
I copied it from `revng_distributable.yml`, dunno if it's needed

aleclearmind replied:
Drop drop drop

aleclearmind commented on .orchestra/support/revng-container/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+COPY apt.no-extra /etc/apt/apt.conf.d/no-extra
+RUN apt update && \
+    apt install -y \
+        # Needed for recompilation pipe (-dev included due to linking problems)
```
Let's be more accurate: we need libc6-dev for `crt*.o` files (which ones?) which are provided by the libc, which we do not install (as anything in `link-only`).

mrjackv replied:
Added comments for each dependency installed so that it's know why they are there

aleclearmind commented on .orchestra/support/revng-container/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+        # Needed for revng.cli etc.
+        python3 ca-certificates \
+        # some LLVM libs are not searched properly
+        libllvm12 libc++1-12 libatomic1 && \
```
Be more specific. These are all bugs, we need to at least have a note on how to fix this.

mrjackv replied:
Ditto wrt the review comment above

aleclearmind commented on .orchestra/support/revng-container/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+    for file in *.so.*; do \
+        SO_FILE=$(echo $file | grep -Po '^.*\.so'); \
+        if [ ! -e $SO_FILE ]; then ln -s $file $SO_FILE; fi \
+    done
```
These are all hacks that will need a more proper fix. Let's do only the strictly necessary set of hacks, so just symlink `librt.so` not all the libs.

mrjackv replied:
Ran a round of strace, seems the only library affected was `librt.so`, symlinked only that one

aleclearmind commented on .orchestra/support/revng-container/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+    rm -rf /var/lib/apt/lists/*
+
+
+FROM BASE AS REVNG_RUN
```
Is all of this just not to install `python3-pip` in the final image? If so, document it.
Is it actually of significant size?

mrjackv replied:
1. Ok
2. There are 2 considerations:
  * Doing it this way guarantees that there are no temporary/leftover files from `pip`
  * From a security standpoint it's better to have less tools in the image, just in case

aleclearmind replied:
2. I think it's a bit overkill, unless you foresee we might add more stuff (which we shouldn't, according to orchestra design). Also consider that once we'll move all our python dependencies *within* orchestra, this will be no longer necessary.
  For security considerations, there's https://bootstrap.pypa.io/get-pip.py. But in general, unless you set `noexec` the file system, you're not adding much protection.

Do what you think is more appropriate, considering that this will vanish once we push Python packages within orchestra.
I'm fine with both approaches given this outlook.

mrjackv replied:
Since we're going to get packages in orchestra soon-ish I've opted to not have an intermediate container and to `pip install` directly in the final container

aleclearmind commented on .orchestra/support/revng-container/apt.no-extra:
```diff
@@ -0,0 +1,4 @@
+APT::Get::Install-Recommends "false";
+APT::Get::Install-Suggests "false";
+APT::Install-Suggests "0";
```
Aren't these defaults?

mrjackv replied:
They are, but I'd keep them just to be sure

aleclearmind commented on .orchestra/support/revng-container/entrypoint.sh:
```diff
@@ -0,0 +1,22 @@
+    done
+
+    export REVNG_NOTIFY_FIFOS="$REVNG_CLOUD_GIT_FIFO:context,$REVNG_CLOUD_S3_FIFO:begin"
+fi
```
* Doc
* Are you sure this belongs here? Is this component `revng-container` or `revng-cloud-container`? If this this works for both situations (i.e., this is useful also for non-cloud situation) this is OK.

mrjackv replied:
This is to be run within `minikube` as well as in cloud, dunno if we want to make the distinction or not.
In case I can add the component `revng-cloud-container` that uses the image generated by `revng-container` and adds this file

aleclearmind replied:
It's fine as long as this still works if one wants to use the container without the cloud.

mrjackv replied:
Slightly reworked, now the names are more generic with docs

aleclearmind commented on .orchestra/support/revng-container/requirements.txt:
```diff
@@ -0,0 +1,12 @@
+pyyaml
+requests
+requests-toolbelt
+xdg
```
Mmmh, can we have this file in `revng`? We should really strive to minimize duplication.

mrjackv replied:
I've exposed the `requirements.txt` in revng so that we can drop this one and use that one instead

aleclearmind commented on .orchestra/support/revng-container/install.sh:
```diff
@@ -0,0 +1,50 @@
+function cleanup {
+  trap - EXIT;
+  if [ -n "${TEST_NAME-}" ]; then
+    podman kill "$TEST_NAME"
```
`podman` is now a dependency, we should install it.

mrjackv replied:
Fixed

aleclearmind commented on .orchestra/support/revng-container/install.sh:
```diff
@@ -0,0 +1,50 @@
+  podman run -d \
+    --name="$TEST_NAME" \
+    -p "127.0.0.10:$TEST_PORT:8000" \
+    --security-opt seccomp=unconfined \
```
Why?

mrjackv replied:
https://medium.com/nttlabs/ubuntu-21-10-and-fedora-35-do-not-work-on-docker-20-10-9-1cd439d9921
This article is about docker, but podman seems to have a similar problem where the seccomp profile is blocking syscalls (in my case it shows up as "can't start new thread")

aleclearmind replied:
Seems like `clone3` is a new syscall that's not listed among those whitelisted by default in older version of podman.
Where are you experiencing this? It seems more a configuration problem on the host machine.

mrjackv replied:
Ok, my bad, it was a stale file on my system. Sorry for the noise.
For completeness I checked which distros shipped with the updated config and it seems all of them except debian bullseye

aleclearmind commented on .orchestra/support/revng-container/install.sh:
```diff
@@ -0,0 +1,50 @@
+
+if [ "$RUN_TESTS" -eq 1 ]; then
+  TEST_NAME="revng-container-test_$UUID"
+  TEST_PORT="$((1024 + RANDOM))"
```
This could produce a number greater than 2**16, I think.

mrjackv replied:
Fixed via the use of unix sockets

aleclearmind commented:
Putting this on hold until the 22.04 branch is in.

aleclearmind commented:
Branch went in, please rebase.

aleclearmind commented:
Rebase!

mrjackv commented:
> Rebase!

Rebased!

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,164 @@
+        `# Needed for Recompile* Steps` \
+        -e "^lib64/llvm/clang-release/bin/llc$" \
+        -e "^lib64/llvm/clang-release/bin/llvm-link$" \
+        `# Revng pipeline libraries` \
```
`s/Revng/rev.ng/` everywhere

mrjackv committed:
Split compilers from their runtime

This commit splits compilers from their runtime, allowing for
revng-distributable to ship only the runtime libraries.
This split affects clang-release and toolchain/host/gcc, both of them
now have a "sister" package, clang-runtime and gcc-runtime respectively,
that only contains the runtime libraries.

mrjackv committed:
Drop building target AMDGPU for clang/llvm

The target was needed when shipping mesa, since that component is no
longer present drop also the target from llvm/clang

mrjackv committed:
Outline pyyaml

pyyaml is required in both `revng-qa` and `revng-python-dependencies`,
outline it in a separate component

mrjackv committed:
python: Upgrade to 3.11

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,183 @@
+cp -a "$ORCHESTRA_DOTDIR/support/revng-distributable/environment" environment
+
+# shellcheck disable=SC2094
+orchestra environment | \
```
```suggestion
orc environment | \
```

mrjackv committed:
ida: avoid polluting cwd

When installing ida, fetch.sh causes the downloaded files to be copied
inside the cwd. This is caused by:
* Not changing to $BUILD_DIR
* Not using --no-copy
Fix both issues in the configure script

mrjackv committed:
Drop revng-qa from revng{,-c}

Drop revng-qa as a dependency for revng and revng-c, since revng-qa is
needed by their relative `test/` package

aleclearmind commented on .orchestra/support/revng-distributable/install.sh:
```diff
@@ -0,0 +1,134 @@
+
+cd "$ORCHESTRA_ROOT"
+
+orc inspect component dependencies --runtime revng-distributable | \
```
Didn't we say to just have `--build-only` to show only build deps? Or is specifying either `--runtime` or `--build` mandatory?

mrjackv replied:
* No switch => both build and runtime
* `--runtime` => only runtime

aleclearmind replied:
Ah right!

aleclearmind commented on .orchestra/support/save-for-later:
```diff
@@ -0,0 +1,56 @@
+            if any(p.match(full_file) for p in patterns):
+                selected_files.append(full_file)
+
+    if not os.path.isdir(SAVE_DIRECTORY):
```
I thought you were a `Pathlib` fag.

mrjackv replied:
If it's a one-off thing I'll use `os.path`, but you're right, I should switch to `pathlib`

aleclearmind commented on .orchestra/config/components/revng.yml:
```diff
@@ -18,8 +18,8 @@ dependencies:
   - boost
-  - clang-release
   - revng-python-dependencies
+  - toolchain/host/binutils
```
This is still WIP, it's part of my branch. I'll do this.

mrjackv replied:
Ok

aleclearmind commented on .orchestra/config/components/clang_runtime.yml:
```diff
@@ -0,0 +1,30 @@
+build_dependencies:
+  - clang-release
+dependencies:
+  - zlib
```
Uh?

mrjackv replied:
I was under the impression that `libc++*` would need zlib since it's a build dependency of `clang-release`, but I was wrong

aleclearmind commented on .orchestra/config/components/gcc_runtime.yml:
```diff
@@ -0,0 +1,24 @@
+
+#@yaml/text-templated-strings
+---
+#@ def gcc_runtime_args():
+license: LICENSE.TXT
+configure: |
+  mkdir -p "$BUILD_DIR"
+install: |
+  cd "$BUILD_DIR"
+  cp -a "${ORCHESTRA_ROOT}/share/orchestra/toolchain_host_gcc.license" LICENSE.TXT
+  tar xf "${ORCHESTRA_ROOT}/share/orchestra/save_for_later/gcc-runtime.tar.xz" \
+    -C "${DESTDIR}${ORCHESTRA_ROOT}"
+build_dependencies:
+  - toolchain/host/gcc
+#@ end
```
We could factor this out, let's keep it in mind for the next time we do `save_for_later`.

mrjackv replied:
Ok!

aleclearmind commented:
Merged in 5a275c8. Thanks!
